src/encord_active/lib/model_predictions/writer.py

```diff
@@ -390,77 +477,86 @@ def add_prediction(
-            # ptype = PredictionType.FRAME
-        elif bbox is None and isinstance(polygon, (np.ndarray, list)):
-            ptype = PredictionType.POLYGON
-            if isinstance(polygon, list):
-                polygon = np.array(polygon)
-
-            if polygon.ndim != 2:
-                raise ValueError("Polygon argument should have just 2 dimensions: [h, w] or [N, 2]")
-
-            if polygon.shape[1] != 2:  # Polygon is mask
-                np_mask = polygon
-                x1, y1, w, h = cv2.boundingRect(polygon)  # type: ignore
-            else:  # Polygon is points
-                # Read image size from label row
-                if np.issubdtype(polygon.dtype, np.integer):
-                    polygon = polygon.astype(float) / np.array([[width, height]])
-
-                np_mask = points_to_mask(polygon, width=width, height=height)  # type: ignore
-                x1, y1, w, h = cv2.boundingRect((polygon * np.array([[width, height]])).reshape(-1, 1, 2).astype(int))  # type: ignore
-            x2, y2 = x1 + w, y1 + h
-            rle = binary_mask_to_rle(np_mask)
-
-        elif isinstance(bbox, dict) and polygon is None:
-            ptype = PredictionType.BBOX
-            self.__check_bbox(bbox)
-            x1 = bbox["x"] * width
-            y1 = bbox["y"] * height
-            x2 = (bbox["x"] + bbox["w"]) * width
-            y2 = (bbox["y"] + bbox["h"]) * height
+        if prediction.classification:
+            class_id = self.classification_class_id_lookup.get(prediction.classification)
+            ptype = PredictionType.FRAME
+            x1, y1, x2, y2 = [None, None, None, None]
+        elif prediction.object:
+            if isinstance(prediction.object.data, np.ndarray):
+                class_id = self.object_class_id_lookup.get(prediction.object.object_class_hash)
+                polygon = prediction.object.data
+                ptype = PredictionType.POLYGON
+                if isinstance(polygon, list):
+                    polygon = np.array(polygon)
+
+                if polygon.ndim != 2:
+                    raise ValueError("Polygon argument should have just 2 dimensions: [h, w] or [N, 2]")
+
+                if polygon.shape[1] != 2:  # Polygon is mask
+                    np_mask = polygon
+                    x1, y1, w, h = cv2.boundingRect(polygon)  # type: ignore
+                else:  # Polygon is points
+                    # Read image size from label row
+                    if np.issubdtype(polygon.dtype, np.integer):
+                        polygon = polygon.astype(float) / np.array([[width, height]])
+
+                    np_mask = points_to_mask(polygon, width=width, height=height)  # type: ignore
+                    x1, y1, w, h = cv2.boundingRect((polygon * np.array([[width, height]])).reshape(-1, 1, 2).astype(int))  # type: ignore
+                x2, y2 = x1 + w, y1 + h
+                rle = binary_mask_to_rle(np_mask)
```

Comment:
I initially wanted to change this and would definitely make it easier to read. However, each branch relies on the current conditions for proper typing. Since I couldn't find a way to tell python the types are matching, we can change `Prediction` to not have `format` and `data` and instead have data of type `Polygon`, `Mask` or `BoundingBox` (but I think this isn't in the scope of this PR).